Charlie

charlie-becker opened the pull request:
Runs a P-type neural network on historical NWP data across the entire grid using interpolated height-above-surface data. Some details:

- It currently supports output for the RAP, HRRR and GFS models. NAM is is also in the config but seems to be buggy.
- Outputs each init time / forecast hour as a separate netCDF file.
-  The only parallelization currently implemented is the downloading of the grib files themselves, which is done through the `FastHerbie()` class. Everything else is done in serial.
- It takes approximately 20 seconds to run the entire RAP grid per forecast hour, but substantially longer for the higher resolution models or global models (~3 minutes).
- Final output has probabilities for each p-type as it's own variable as well as the predicted class (argmax) for each of the p-types. It also saves out surface variables `['t2m', 'd2m', 'u10', 'v10' , 'orog']`, but does not save out the pressure level data unless flagged in the configuaration file. 

Future features:

- Parallelize / optimize code to reduce time for higher resolution models.
- Add feature to only run over a specific extent. This will still require the download of the entire file, but will significantly speed up the entire process due to the amount of data that needs to be interpolated, etc.
- Add ability to run in real time. Could be done with the `Herbie_latest()` class or manually with a small lag from script run time.
- Currently only supports a neural network with an output size of 4. New feature should include ability to account for models with uncertainty.
- Feature to save out the categorical p-type from the NWP model.

Output seems reasonable, but I would like someone to review the interpolation code `interpolate_data()` in `inference.py` as a sanity check.

Commit:
Initial inference commit (config / script)

Commit:
Update inference config

Commit:
Add inference.py as primary inference module

Commit:
Add run script

Commit:
update requirements and environment

Commit:
update path

Commit:
update config

charlie-becker committed:
Merge branch 'main' into charlie

Commit:
remove old inference script (renamed)

charlie-becker committed:
remove old inference.py (wrong loaction / renamed)

Commit:
update config

Commit:
Merge remote-tracking branch 'origin/charlie' into charlie

Commit:
update requirements

Commit:
Change to include NWP P-type

Commit:
update config for NWP p-type

Commit:
Some optimization (numba / numpy)

djgagne commented on ptype/inference.py:
```diff
@@ -0,0 +1,260 @@
+    return pd.DataFrame(flat_data, columns=cols)
+
+
+def kelvin_to_celcius(temp):
```
The temperature is spelled "Celsius".

djgagne commented on ptype/inference.py:
```diff
@@ -0,0 +1,260 @@
+    date_str = date.strftime(f"%Y%m%d_{forecast_hour:02}00")
+    file_str = f"ptype_predictions_{model}_{date_str}.nc"
+    full_path = os.path.join(out_path, model, dir_str, file_str)
+    dataset.to_netcdf(full_path)
```
This saves an uncompressed netcdf file, which can be really large in many cases. Please add options to do lossless and/or lossy compression with the encoding kwarg.

charlie-becker replied:
Resolved

djgagne commented on ptype/inference.py:
```diff
@@ -0,0 +1,260 @@
+
+    model = CategoricalDNN(**conf["model"])
+    model.build_neural_network(268, 4)
+    model.model.load_weights(os.path.join(model_path, "models", "model_11.h5"))
```
If you use keras.models.load_model, you might not have to build a neural network before loading it. Also, please do not hard code model names and input sizes and whatnot. Make them arguments to the function.

djgagne commented on ptype/inference.py:
```diff
@@ -0,0 +1,260 @@
+@jit(nopython=True, parallel=True, cache=True)
+def interpolate(x, y, height_levels):
+
+    arr = np.zeros(shape=(x.shape[0], len(height_levels)))
```
I suggest using more descriptive variable names like height_interp_data.

djgagne commented on ptype/inference.py:
```diff
@@ -0,0 +1,260 @@
+                    "filter_by_keys": {'typeOfLevel': key, 'shortName': var, 'stepType': 'instant'}})
+            grib_data.append(grib)
+
+    ds = xr.merge(grib_data, compat='override')#.load()
```
Can you use more robust variable names to describe the different parts of the pipeline. ds and df are really generic, so it can be hard to track their functionality.

djgagne commented on scripts/run_inference.py:
```diff
@@ -0,0 +1,72 @@
+    parser.add_argument("-c", "--config", help="Path to config file")
+    args = parser.parse_args()
+    # with open(args.config) as config_file:
+    with open("/Users/cbecker/Desktop/Projects/ptype-physical/config/inference.yml") as config_file:
```
Please uncomment args.config so this function can be used by other people.

djgagne reviewed (changes requested):
The general workflow looks correct, and I am glad it is running really fast. I have a number of style/formatting/variable name comments that need to be addressed before merging.

Commit:
deletion of index files

Commit:
Spelling mistake

Commit:
Compress netCDF files

Commit:
remove index / grib files

Commit:
Change variable names for clarity

Commit:
remove hard code of model size

Commit:
change batch size

Commit:
minor change

charlie-becker committed:
Update run_inference.py args

charlie-becker commented:
All requested changes have been addressed other than the load model. I have not been able to successfully get `keras.models.load_model()` to work. I'll work with @jsschreck to see if that can be resolved.

@djgagne I have also successfully parallelized this code with both multi-processing and dask (dask for large historical runs, and multiprocessing for real-time). Would you like me to add that to this PR? Or should we merge and create a new one?

djgagne commented on ptype/inference.py:
```diff
@@ -187,11 +196,13 @@ def load_model(model_path):
     model = CategoricalDNN(**conf["model"])
-    model.build_neural_network(268, 4)
+    model.build_neural_network(len(x_transformer.x_columns_), len(output_scaler['classes_']))
     model.model.load_weights(os.path.join(model_path, "models", "model_11.h5"))
```
Can you make sure the model filename and scaler filenames are also not hard coded?

Commit:
Parallelization

Commit:
uypdate config for parallelization

Commit:
add NWP ptype categorical variables to output

Commit:
update environment and requirements

Commit:
Add inference portion to README documentation

Commit:
Merge remote-tracking branch 'origin/charlie' into charlie

charlie-becker commented:
Added code for parallelization, adding NWP categorical variables, and updated documentation.

djgagne commented on ptype/inference.py:
```diff
@@ -232,29 +246,17 @@ def save_data(dataset, out_path, date, model, forecast_hour):
+    full_path = os.path.join(out_path, model, dir_str, model_run_str, file_str)
+    encoding_vars = [v for v in list(dataset.data_vars)]
+    encoding = {var: {"zlib": True, "complevel": 4, "least_significant_digit": 4} for var in encoding_vars}
+    dataset.to_netcdf(full_path, encoding=encoding)
```
Can you add a to_zarr option here as well? The Vaisala team was interested in Zarr output since they operate on AWS and it would be easier for them to download.

Commit:
add option to save in zarr format

Commit:
add file names (model, scalers)

Commit:
add compression for zarr

charlie-becker commented:
All comments resolved. Some chunking optimization for zarr should be investigated at some point.

Commit:
add proj coords, change file names and dtpyes

charlie-becker commented:
* Added projection attribute and projection coordinates
* Changed file naming structure to **MILES_ptype_{nwp_model}_YYYY-mm-dd_MMHH_f{forecast_hour}.zarr** (or netcdf)
* Changed dtypes to float32 and uint8
* Added time dimension (size 1) for easier concatenation

Here is some example output for the HRRR model:


 
![Screen Shot 2023-05-17 at 3 57 01 PM](https://github.com/ai2es/ptype-physical/assets/31113893/32d7f598-00a5-410c-b754-a3567d1c6f9a)

charlie-becker committed:
Small dask arg bug

djgagne commented on ptype/inference.py:
```diff
@@ -0,0 +1,319 @@
+    return model, x_transformer
+
+
+def grid_preditions(data, preds):
```
spelling error. Should be grid_predictions

charlie-becker replied:
resolved

charlie-becker committed:
Fix spelling error

charlie-becker committed:
fix spelling error

charlie-becker commented:
Merging.
